pkg/provider/core.go

```diff
@@ -0,0 +1,388 @@
+	}
+	device, _, err := svc.Devices.Get(id, &packngo.GetOptions{})
+	if err != nil {
+		klog.Errorf("Could not get device %s: %v", id, err)
+		return nil, err
+	}
+
+	klog.V(2).Infof("Machine get request has been processed successfully for %q", name)
+	return &driver.GetMachineStatusResponse{
+		NodeName:   name,
+		ProviderID: encodeMachineID(device.Facility.String(), device.ID),
+	}, nil
+}
+
+// ListMachines lists all the machines possibilly created by a providerSpec
+// Identifying machines created by a given providerSpec depends on the OPTIONAL IMPLEMENTATION LOGIC
+// you have used to identify machines created by a providerSpec. It could be tags/resource-groups etc
+// OPTIONAL METHOD
+//
+// REQUEST PARAMETERS (driver.ListMachinesRequest)
+// MachineClass          *v1alpha1.MachineClass   MachineClass based on which VMs created have to be listed
+// Secret                *corev1.Secret           Kubernetes secret that contains any sensitive data/credentials
+//
+// RESPONSE PARAMETERS (driver.ListMachinesResponse)
+// MachineList           map<string,string>  A map containing the keys as the MachineID and value as the MachineName
+//                                           for all machine's who where possibilly created by this ProviderSpec
+//
+func (p *Provider) ListMachines(ctx context.Context, req *driver.ListMachinesRequest) (*driver.ListMachinesResponse, error) {
+	// Log messages to track start and end of request
+	klog.V(2).Infof("List machines request has been received for %q", req.MachineClass.Name)
+
+	var (
+		clusterName, nodeRole string
+		resp                  = &driver.ListMachinesResponse{
+			MachineList: make(map[string]string),
+		}
+	)
+
+	// Check if incoming CR is a CR we support
+	if req.MachineClass.Provider != ProviderEquinixMetal {
+		return nil, fmt.Errorf("Requested for Provider '%s', we only support '%s'", req.MachineClass.Provider, ProviderEquinixMetal)
+	}
+
+	providerSpec, err := decodeProviderSpecAndSecret(req.MachineClass, req.Secret)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, key := range providerSpec.Tags {
+		if strings.Contains(key, "kubernetes.io/cluster/") {
+			clusterName = key
+		} else if strings.Contains(key, "kubernetes.io/role/") {
+			nodeRole = key
+		}
+	}
+
+	if clusterName == "" || nodeRole == "" {
+		return resp, nil
+	}
+
+	svc := p.createSVC(getApiKey(req.Secret))
+	if svc == nil {
+		return nil, fmt.Errorf("nil Equinix Metal service returned")
+	}
+	devices, _, err := svc.Devices.List(providerSpec.ProjectID, &packngo.ListOptions{})
+	if err != nil {
+		klog.Errorf("Could not list devices for project %s: %v", providerSpec.ProjectID, err)
+		return nil, err
+	}
+	for _, d := range devices {
+		matchedCluster := false
+		matchedRole := false
+		for _, tag := range d.Tags {
+			switch tag {
+			case clusterName:
+				matchedCluster = true
+			case nodeRole:
+				matchedRole = true
+			}
+		}
+		if matchedCluster && matchedRole {
+			resp.MachineList[d.ID] = d.Hostname
```

Comment:
We may have to encodeMachineID for the d.ID here again. @prashanth26 Please confirm my doubt again

Reply:
Done